src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/AspNet/ApiController_ExecuteAsync_Integration.cs

```diff
@@ -0,0 +1,124 @@
+#if NETFRAMEWORK
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Web;
+using Datadog.Trace.ClrProfiler.CallTarget;
+using Datadog.Trace.ClrProfiler.Emit;
+using Datadog.Trace.ClrProfiler.Integrations;
+using Datadog.Trace.ClrProfiler.Integrations.AspNet;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.ExtensionMethods;
+using Datadog.Trace.Tagging;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.AspNet
+{
+    /// <summary>
+    /// System.Web.Http.ApiController.ExecuteAsync calltarget instrumentation
+    /// </summary>
+    [InstrumentMethod(
+        AssemblyName = SystemWebHttpAssemblyName,
+        TypeName = "System.Web.Http.ApiController",
+        MethodName = "ExecuteAsync",
+        ReturnTypeName = ClrNames.HttpResponseMessageTask,
+        ParameterTypeNames = new[] { HttpControllerContextTypeName, ClrNames.CancellationToken },
+        MinimumVersion = Major5Minor1,
+        MaximumVersion = Major5MinorX,
+        IntegrationName = IntegrationName)]
+    // ReSharper disable once InconsistentNaming
+    public class ApiController_ExecuteAsync_Integration
+    {
+        private const string SystemWebHttpAssemblyName = "System.Web.Http";
+        private const string HttpControllerContextTypeName = "System.Web.Http.Controllers.HttpControllerContext";
+        private const string Major5Minor1 = "5.1";
+        private const string Major5MinorX = "5";
+
+        private const string IntegrationName = nameof(IntegrationIds.AspNetWebApi2);
+
+        /// <summary>
+        /// OnMethodBegin callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <typeparam name="TController">Type of the controller context</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <param name="controllerContext">The context of the controller</param>
+        /// <param name="cancellationToken">The cancellation token</param>
+        /// <returns>Calltarget state value</returns>
+        public static CallTargetState OnMethodBegin<TTarget, TController>(TTarget instance, TController controllerContext, CancellationToken cancellationToken)
+            where TController : IHttpControllerContext
+        {
+            var scope = AspNetWebApi2Integration.CreateScope(controllerContext, out _);
+
+            if (scope != null)
+            {
+                return new CallTargetState(scope, controllerContext);
+            }
```

Comment:
I think in this case we lose all the benefits of using the `constraint` trick, the `struct` created by the ducktyping will be boxed in both `CreateScope` and `new `CallTargetState`.

So for this I rather prefer to remove the `constraints` and convert the `IHttpControllerContext` to a struct with the `[DuckCopy]` attribute, and just use the `As<T>` method when we need the access, that will prevent the additional allocation.

Or...

If we accept that single allocation then do the boxing once at the beginning of the method?

Reply:
As discussed on Slack, I'll go for the "box only once" way